programs/ssl/ssl_base64_dump.c

```diff
@@ -0,0 +1,953 @@
+            return "none";
+        case MBEDTLS_SSL_MAX_FRAG_LEN_512:
+            return "512";
+        case MBEDTLS_SSL_MAX_FRAG_LEN_1024:
+            return "1024";
+        case MBEDTLS_SSL_MAX_FRAG_LEN_2048:
+            return "2048";
+        case MBEDTLS_SSL_MAX_FRAG_LEN_4096:
+            return "4096";
+        default:
+            return "error";
+    }
+}
+
+/*
+ * Read next base64 code form the 'b64_file'. The 'b64_file' must be opened
+ * previously. After next call of this function, the search will take from the
+ * place where it was previously end.
+ *
+ * /p b64       buffer for input data
+ * /p max_len   the maximum number of bytest to write
+ *
+ * \retval      number of bytes writen in to the b64 buffer or 0 in case no more
+ *              data was found
+ */
+size_t read_next_b64_code( uint8_t *b64, size_t max_len )
+{
+    size_t len = 0;
+    size_t missed = 0;
+    char pad = 0;
+    char c = 0;
+
+    while( EOF != c )
+    {
+        char c_valid = 0;
+
+        c = (char)fgetc( b64_file );
+
+        if( pad == 1 )
+        {
+            if( c == '=' )
+            {
+                c_valid = 1;
+                pad = 2;
+            }
+        }
+        else if( ( c >= 'A' && c <= 'Z' ) ||
+                 ( c >= 'a' && c <= 'z' ) ||
+                 ( c >= '0' && c <= '9' ) ||
+                   c == '+' || c == '/' )
+        {
+            c_valid = 1;
+        }
+        else if( c == '=' )
+        {
+            c_valid = 1;
+            pad = 1;
+        }
+        else if( c == '-' )
+        {
+            c = '+';
+            c_valid = 1;
+        }
+        else if( c == '_' )
+        {
+            c = '/';
+            c_valid = 1;
+        }
+
+        if( c_valid )
+        {
+            if( len < max_len )
+            {
+                b64[ len++ ] = c;
+            }
+            else
+            {
+                missed++;
+            }
+        }
+        else if( len > 0 )
+        {
+            if( missed > 0 )
+            {
+                printf_err( "Buffer for the base64 code is too low. Missed %zu characters\n", missed );
+            }
+            return len;
+        }
+    }
+
+    printf_dbg( "End of file\n" );
+    return 0;
+}
+
+/*
+ * This function deserialize and print at the stdout all obtained information
+ * about the certificates from provided data.
+ *
+ * /p ssl   pointer to serialized certificate
+ * /p len   number of bytes in the buffer
+*/
+void print_deserialized_ssl_cert( const uint8_t *ssl, size_t len )
+{
+    enum { STRLEN = 4096 };
+    mbedtls_x509_crt crt;
+    int ret;
+    char str[STRLEN];
+
+    printf( "\nCertificate:\n" );
+
+    mbedtls_x509_crt_init( &crt );
+    ret = mbedtls_x509_crt_parse_der( &crt, ssl, len );
+    if( 0 != ret )
+    {
+        mbedtls_strerror( ret, str, STRLEN );
+        printf_err( "Invalid format of X.509 - %s\n", str );
+        printf( "Cannot deserialize:\n\t" );
```

Comment:
Isn't this an error?

Reply:
It was intended to use `printf` here instead of `printf_err` to avoid two "ERROR: " messages in two lines